### PR TITLE
Porting libnetwork fixes

### PR DESCRIPTION
**Pull request by aboch**
Brings to 1.12.x the fixes for the following issues:

- #28112
- #28465
- #28606 
- #28630 
- #25184 
-  #28697 
- #28845 
- #27188 
- #28055 
- #26251 
-  docker/for-mac#849
- docker/for-mac/issues/888
- docker/libnetwork/issues/1474

### CODE REVIEW

**Comment by vieux** [comment]
@aboch looks like issues

**Comment by aboch** [comment]
Not sure about vendor failure:
```
00:33:28 hack/dind: line 29: /go/src/github.com/docker/docker/hack/validate/vendor: No such file or directory
00:33:29 Build step 'Execute shell' marked build as failure
```
It looks like it is looking for the `validate` directory that was added for the new vendoring, which is not present in 1.12.x branch.

**Comment by anusha-ragunathan** [comment]
experimental is failing because in 1.12, plugins are still in experimental. 

The media type for plugins was changed in 1.13 to `application/vnd.docker.plugin.v1+json` https://github.com/docker/distribution/blob/4ad885a6725bfca9ea00bd87341e367c0f919560/manifest/schema2/manifest.go#L21. As a result, the sample plugin we use in integration tests `tiborvass/no-remove` was re-pushed to Hub using this new mediatype.

That's why the tests are failing. We have 2 options:
1. Backport the plugin mediatype changes to 1.12.x
2. Repush a 1.12 version of `tiborvass/no-remove` that has the old mediatype and update the tests to use it.

I prefer (1)

**Comment by anusha-ragunathan** [comment]
For the plugin mediatype changes, we need https://github.com/docker/docker/commit/970b23db2e97e1671c6bf4b57bfe5d1a30e98819

**Comment by anusha-ragunathan** [comment]
An easier option is to simply add the new media type here https://github.com/aboch/docker/blob/76d2f1f29bbc11c92ab20489fc5eee2593b6d879/plugin/distribution/types.go#L16. Then use both the new and the old type for checking in https://github.com/aboch/docker/blob/76d2f1f29bbc11c92ab20489fc5eee2593b6d879/plugin/distribution/pull.go#L146

**Comment by vieux** [comment]
@johnstep can you please try this one, since there is no CI for 1.12.x on Windows

**Comment by johnstep** [comment]
LGTM on Windows. It built successfully on Windows (for Windows) and all unit tests passed. I rebuilt and tested again with the resulting binaries, and that succeeded.

File: plugin/distribution/pull.go [hunk 1]
```diff
@@ -143,7 +143,8 @@ func Pull(name string, rs registry.Service, metaheader http.Header, authConfig *
 		logrus.Debugf("pull.go: error in json.Unmarshal(): %v", err)
 		return nil, err
 	}
-	if m.Config.MediaType != MediaTypeConfig {
+	if m.Config.MediaType != MediaTypeConfig &&
+		m.Config.MediaType != "application/vnd.docker.plugin.v0+json" {
```
Comment (anusha-ragunathan) [inline plugin/distribution/pull.go]: If this check is updated as follows, we should be good:

```
if m.Config.MediaType != MediaTypeConfig &&
   m.Config.MediaType != "application/vnd.docker.plugin.v0+json" &&
   m.Config.MediaType != "application/vnd.docker.plugin.image.v0+json"{
...
}
```

**Comment by anusha-ragunathan** [comment]
All experimental tests are failing due to the known plugin mediatype issue. It can be fixed by updating the code to accommodate to the expected media type of the sample plugin. Or the sample plugin can be re-pushed by @tiborvass 

In either case, this PR can be merged, since the failing tests are unrelated to the PR.

**Comment by vieux** [comment]
LGTM ping @LK4D4 @tonistiigi

**Comment by aboch** [comment]
@anusha-ragunathan Thanks for your help in trying to solve the experimental failure on 1.12.x.
Given it is currently broken, I feel it would be more efficient if you guys address the plugin failure on experimental ci issue via a separate PR. Will also give you more control over what to change. After that is merged we can rebase our vendoring (libnetwork/swarmkit) PRs .
I went ahead and removed the uncomplete media type changes commit from this PR.

**Comment by anusha-ragunathan** [comment]
@aboch : Yessir! https://github.com/docker/docker/pull/29089

**Comment by vieux** [comment]
@tonistiigi windows is expected to fail

**Comment by mavenugo** [comment]
#29089 handles the experimental failures.

LGTM

**Pull request merged**